dask_expr/merge.py

```diff
@@ -0,0 +1,177 @@
+
+    _parameters = [
+        "left",
+        "right",
+        "how",
+        "left_on",
+        "right_on",
+        "left_index",
+        "right_index",
+        "suffixes",
+        "indicator",
+        "shuffle_backend",
+    ]
+    _defaults = {
+        "how": "inner",
+        "left_on": None,
+        "right_on": None,
+        "left_index": False,
+        "right_index": False,
+        "suffixes": ("_x", "_y"),
+        "indicator": False,
+        "shuffle_backend": None,
+    }
+
+    def __str__(self):
+        return f"Merge({self._name[-7:]})"
+
+    @property
+    def kwargs(self):
+        return {
+            k: self.operand(k)
+            for k in [
+                "how",
+                "left_on",
+                "right_on",
+                "left_index",
+                "right_index",
+                "suffixes",
+                "indicator",
+            ]
+        }
+
+    @functools.cached_property
+    def _meta(self):
+        left = meta_nonempty(self.left._meta)
+        right = meta_nonempty(self.right._meta)
+        return make_meta(left.merge(right, **self.kwargs))
+
+    def _divisions(self):
+        npartitions_left = self.left.npartitions
+        npartitions_right = self.right.npartitions
+        npartitions = max(npartitions_left, npartitions_right)
+        return (None,) * (npartitions + 1)
+
+    def _lower(self):
+        # Lower from an abstract expression using
+        # logic in MergeBackend.from_abstract_merge
+
+        left = self.left
+        right = self.right
+        how = self.how
+        left_on = self.left_on
+        right_on = self.right_on
+        left_index = self.left_index
+        right_index = self.right_index
+        shuffle_backend = self.shuffle_backend
+
+        # TODO:
+        #  1. Handle mixed indexed merge
+        #  2. Add multi-partition broadcast merge
+        #  3. Add/leverage partition statistics
+
+        # Check for "trivial" broadcast (single partition)
+        npartitions = max(left.npartitions, right.npartitions)
+        if (
+            npartitions == 1
+            or left.npartitions == 1
+            and how in ("right", "inner")
+            or right.npartitions == 1
+            and how in ("left", "inner")
+        ):
+            return BlockwiseMerge(left, right, **self.kwargs)
+
+        # Check if we are merging on indices with known divisions
+        merge_indexed_left = (
+            left_index or _contains_index_name(left, left_on)
+        ) and left.known_divisions
+        merge_indexed_right = (
+            right_index or _contains_index_name(right, right_on)
+        ) and right.known_divisions
+
+        # NOTE: Merging on an index is fragile. Pandas behavior
+        # depends on the actual data, and so we cannot use `meta`
+        # to accurately predict the output columns. Once general
+        # partition statistics are available, it may make sense
+        # to drop support for left_on and right_on.
```

Comment:
The only component of this PR that doesn't feel "ready" is that we aren't really testing the case that we want to merge on an index, but require a shuffle. However, while trying to add a test for this, I discovered that merging on an index is actually quite fragile in general :/

Reply:
oops `left_on and right_on` should be `left_index and right_index` here.